pkg/utils/permission.go

```diff
@@ -0,0 +1,275 @@
+			if err != nil {
+				return err
+			}
+		} else {
+			clusterRole.ResourceVersion = latest.ResourceVersion
+			_, err := p.kubeClient.RbacV1().ClusterRoles().Update(context.TODO(), clusterRole, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	return p
+}
+
+func (p *permissionBuilder) WithStaticClusterRoleBinding(binding *rbacv1.ClusterRoleBinding) RBACPermissionBuilder {
+	p.u.fns = append(p.u.fns, func(cluster *clusterv1.ManagedCluster, addon *addonapiv1alpha1.ManagedClusterAddOn) error {
+		ensureAddonOwnerReference(&binding.ObjectMeta, addon)
+		exists := true
+		latest, err := p.kubeClient.RbacV1().ClusterRoleBindings().Get(context.TODO(), binding.Name, metav1.GetOptions{})
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			exists = false
+		}
+		if !exists {
+			_, err := p.kubeClient.RbacV1().ClusterRoleBindings().Create(context.TODO(), binding, metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
+		} else {
+			binding.ResourceVersion = latest.ResourceVersion
+			_, err := p.kubeClient.RbacV1().ClusterRoleBindings().Update(context.TODO(), binding, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	return p
+}
+
+func (p *permissionBuilder) WithStaticRole(role *rbacv1.Role) RBACPermissionBuilder {
+	p.u.fns = append(p.u.fns, func(cluster *clusterv1.ManagedCluster, addon *addonapiv1alpha1.ManagedClusterAddOn) error {
+		role.Namespace = cluster.Name
+		ensureAddonOwnerReference(&role.ObjectMeta, addon)
+		exists := true
+		latest, err := p.kubeClient.RbacV1().Roles(cluster.Name).Get(context.TODO(), role.Name, metav1.GetOptions{})
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			exists = false
+		}
+		if !exists {
+			_, err := p.kubeClient.RbacV1().Roles(cluster.Name).Create(context.TODO(), role, metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
+		} else {
+			role.ResourceVersion = latest.ResourceVersion
+			_, err := p.kubeClient.RbacV1().Roles(cluster.Name).Update(context.TODO(), role, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	return p
+}
+
+func (p *permissionBuilder) WithStaticRoleBinding(binding *rbacv1.RoleBinding) RBACPermissionBuilder {
+	p.u.fns = append(p.u.fns, func(cluster *clusterv1.ManagedCluster, addon *addonapiv1alpha1.ManagedClusterAddOn) error {
+		binding.Namespace = cluster.Name
+		ensureAddonOwnerReference(&binding.ObjectMeta, addon)
+		exists := true
+		latest, err := p.kubeClient.RbacV1().RoleBindings(cluster.Name).Get(context.TODO(), binding.Name, metav1.GetOptions{})
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			exists = false
+		}
+		if !exists {
+			_, err := p.kubeClient.RbacV1().RoleBindings(cluster.Name).Create(context.TODO(), binding, metav1.CreateOptions{})
```

Comment:
use ApplyRole in library-go? I think we need to handle the rule change on rbac

Reply:
good point, replace w `ApplyXXX`